src/networking.c

```diff
@@ -2510,8 +2517,18 @@ int processInputBuffer(client *c) {
         }
     }
 
-    /* Trim to pos */
-    if (c->qb_pos) {
+    if (c->flags & CLIENT_MASTER) {
+        /* The master client is very special, since the querybuf contains data not only
+         * read from network, but also part of replication stream. So we should propagate
+         * the applied data and only trim the applied part of querybuf. */
+        if (c->repl_applied) {
+            replicationFeedStreamFromMasterStream(c->querybuf,c->repl_applied);
```

Comment:
since we feed the stream here, in theory we can completely avoid feeding the stream in processPendingCommandsAndResetClient or commandProcessed.
but one problem with this would be that if some commands executed successfully, and then the client dropped, this block will not be reached, so either we change the `return` above to `break`, or maybe better yet, do the feeding in commandProcessed, remember the offsets, and then do the trimming (sdsrange) here.

Reply:
>but one problem with this would be that if some commands executed successfully, and then the client dropped, this block will not be reached, so either we change the `return` above to `break`

I thought about it, and I think it would not happen, the scenario you said is, master client dropped when `processCommandAndResetClient`, in this function only two conditions lead to client dropped:
1. `evictClients` drop itself.
2. `performEvictions` meets network error when send data to replica and the current client is the replica.

so a master client would not be dropped here, and I add `serverAssert(server.master->repl_applied == 0);` when drop master client.

>or maybe better yet, do the feeding in commandProcessed, remember the offsets, and then do the trimming (sdsrange) here.

your suggestion is good, I'm thinking about it.